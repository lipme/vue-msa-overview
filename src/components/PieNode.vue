<template>
<g :class="classAttribute" :transform="transformAttribute" style="transition:transform 0.5s ease-out">
     <path v-for="(d, i) in pieData" :key="d.data.label" :d="paths[i]" :fill="d.data.color">
        <title>{{d.data.label}} : {{d.data.value}}</title>
     </path>
    </g>
</template>

<script>
import * as d3 from 'd3'

export default {
  name: 'pieNode',
  props: {
    size: {
      type: Number,
      default: 6
    },
    x: {
      type: Number,
      default: 0
    },
    y: {
      type: Number,
      default: 0
    },
    circular: {
      type: Boolean,
      default: false
    },
    data: {
      type: Array,
      default: () => { return [] }
    }
  },

  computed: {
    strokeWidth () {
      return this.size / 5
    },
    classAttribute () {
      return 'pie ' + this.type
    },
    transformAttribute () {
      if (this.circular === false) {
        return 'translate(' + this.y + ' ' + this.x + ')'
      } else {
        return 'rotate(' + (this.x - 90) + ')translate(' + this.y + ')'
      }
    },
    arcFunction () {
      return d3.arc().innerRadius(0).outerRadius(this.size / 2)
    },
    pieData () {
      const pie = d3.pie().value(d => d.value).sort(null)
      return pie(this.data)
    },
    paths () {
      const arc = this.arcFunction
      return this.pieData.map(d => { return arc(d) })
    }
  }
}
</script>

<style scoped>
path:hover {
  transition: all 0.5s;
  transform: scale(1.5);
}
</style>
