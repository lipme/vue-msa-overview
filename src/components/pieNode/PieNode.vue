// Copyright 2020 [LIPM]
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

// http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
//    limitations under the License.
<template>
<g v-if="data.length>0" :transform="transformAttribute">
     <path v-for="(d, i) in pieData" :key="d.data.label" :d="paths[i]" :fill="d.data.color">
        <title>{{d.data.label}} : {{d.data.value}}</title>
        <animateTransform attributeName="transform"
             type="scale"
             from="0 0"
             to="1 1"
             begin="0s"
             dur="1s"
        />
     </path>
    </g>
</template>

<script>
import * as d3 from 'd3'

export default {
  name: 'pieNode',
  props: {
    size: {
      type: Number,
      default: 6
    },
    x: {
      type: Number,
      default: 0
    },
    y: {
      type: Number,
      default: 0
    },
    circular: {
      type: Boolean,
      default: false
    },
    data: {
      type: Array,
      default: () => { return [] },
      required: true,
      validator: function (data) {
        if (data.length === 0) { return false }
        for (let i = 0; i < data.length; i++) {
          if (!('value' in data[i])) {
            console.error("missing key 'value' in pie data")
            return false
          }
          if (!('color' in data[i])) {
            console.error("missing key 'color' in pie data")
            return false
          }
          if (!('label' in data[i])) {
            console.error("missing key 'label' in pie data")
            return false
          }
          if (!isFinite(data[i].value)) {
            console.error("The 'value' attribute in pie data must be a number")
            return false
          }
        }
        return true
      }
    }
  },

  computed: {
    transformAttribute () {
      if (this.circular === false) {
        return 'translate(' + this.y + ' ' + this.x + ')'
      } else {
        return 'rotate(' + (this.x - 90) + ')translate(' + this.y + ')'
      }
    },
    arcFunction () {
      return d3.arc().innerRadius(0).outerRadius(this.size / 2)
    },
    pieData () {
      const pie = d3.pie().value(d => d.value).sort(null)
      return pie(this.data)
    },
    paths () {
      const arc = this.arcFunction
      return this.pieData.map(d => { return arc(d) })
    }
  }
}
</script>

<style scoped>
path:hover {
  transition: all 0.5s;
  transform: scale(1.5);
}
</style>
