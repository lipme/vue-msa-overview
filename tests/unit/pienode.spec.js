import { shallowMount } from '@vue/test-utils'
import PieNode from '@/components/PieNode.vue'

const validator = PieNode.props.data.validator

describe('PieNode component', () => {
  it('does not render if no data', () => {
    const comp = shallowMount(PieNode, {
      propsData: {
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('data set correctly', () => {
    const comp = shallowMount(PieNode, {
      propsData: {
        data: [{ value: 10, color: 'green', label: 'value 1' }]
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('error when value key not set', () => {
    expect(validator([{ color: 'green', label: 'value 1' }])).toBe(false)
  })
  it('error when color key not set', () => {
    expect(validator([{ value: 10, label: 'value 1' }])).toBe(false)
  })
  it('error when label key not set', () => {
    expect(validator([{ color: 'green', value: 10 }])).toBe(false)
  })
  it('error when value key is not a number', () => {
    expect(validator([{ value: 'tuc', color: 'green', label: 'value 1' }])).toBe(false)
  })
  it('size set correctly', () => {
    const comp = shallowMount(PieNode, {
      propsData: {
        data: [{ value: 10, color: 'green', label: 'value 1' }],
        size: 1
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('x set correctly', () => {
    const comp = shallowMount(PieNode, {
      propsData: {
        data: [{ value: 10, color: 'green', label: 'value 1' }],
        x: 1
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('y set correctly', () => {
    const comp = shallowMount(PieNode, {
      propsData: {
        data: [{ value: 10, color: 'green', label: 'value 1' }],
        y: 1
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('circular mode set correctly', () => {
    const comp = shallowMount(PieNode, {
      propsData: {
        data: [{ value: 10, color: 'green', label: 'value 1' }],
        circular: true
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
})
