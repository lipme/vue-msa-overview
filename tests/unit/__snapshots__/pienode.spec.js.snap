// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieNode component circular mode set correctly 1`] = `
<g
  transform="rotate(-90)translate(0)"
>
  <path
    d="M1.8369701987210297e-16,-3A3,3,0,1,1,-1.8369701987210297e-16,3A3,3,0,1,1,1.8369701987210297e-16,-3Z"
    fill="green"
  >
    <title>
      value 1 : 10
    </title>
     
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </path>
</g>
`;

exports[`PieNode component data set correctly 1`] = `
<g
  transform="translate(0 0)"
>
  <path
    d="M1.8369701987210297e-16,-3A3,3,0,1,1,-1.8369701987210297e-16,3A3,3,0,1,1,1.8369701987210297e-16,-3Z"
    fill="green"
  >
    <title>
      value 1 : 10
    </title>
     
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </path>
</g>
`;

exports[`PieNode component does not render if no data 1`] = `<!---->`;

exports[`PieNode component size set correctly 1`] = `
<g
  transform="translate(0 0)"
>
  <path
    d="M3.061616997868383e-17,-0.5A0.5,0.5,0,1,1,-3.061616997868383e-17,0.5A0.5,0.5,0,1,1,3.061616997868383e-17,-0.5Z"
    fill="green"
  >
    <title>
      value 1 : 10
    </title>
     
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </path>
</g>
`;

exports[`PieNode component x set correctly 1`] = `
<g
  transform="translate(0 1)"
>
  <path
    d="M1.8369701987210297e-16,-3A3,3,0,1,1,-1.8369701987210297e-16,3A3,3,0,1,1,1.8369701987210297e-16,-3Z"
    fill="green"
  >
    <title>
      value 1 : 10
    </title>
     
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </path>
</g>
`;

exports[`PieNode component y set correctly 1`] = `
<g
  transform="translate(1 0)"
>
  <path
    d="M1.8369701987210297e-16,-3A3,3,0,1,1,-1.8369701987210297e-16,3A3,3,0,1,1,1.8369701987210297e-16,-3Z"
    fill="green"
  >
    <title>
      value 1 : 10
    </title>
     
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </path>
</g>
`;
